train/train_process.py

```diff
@@ -173,103 +176,144 @@ def _self_play(self, model_path):
-        history = model.fit(train_dataset, epochs=1, verbose=0).history
-        self.metrics.update(
-            {
-                "value_loss": history["value_head_loss"][0],
-                "policy_loss": history["policy_head_loss"][0],
-                "value_accuracy": history["value_head_value_head_accuracy"][0],
-                "policy_accuracy": history["policy_head_policy_head_accuracy"][0],
-                "training_duration": time.time() - training_start_time,
-            }
-        )
-
-        logging.info("Value loss {:.4f}".format(self.metrics["value_loss"]))
-        logging.info("Policy loss {:.4f}".format(self.metrics["policy_loss"]))
-        logging.info("Value accuracy {:.4f}".format(self.metrics["value_accuracy"]))
-        logging.info("Policy accuracy {:.4f}".format(self.metrics["policy_accuracy"]))
-
-        return self._save_model(model)
-
-    def _compare_models(self, best_model, latest_model):
+        logging.debug("Training models with learning rate %f", lr)
+
+        values_loss = [None] * len(models)
+        policy_loss = [None] * len(models)
+        value_accuracy = [None] * len(models)
+        policy_accuracy = [None] * len(models)
+        training_duration = [None] * len(models)
+        trained_models = [None] * len(models)
+
+        def train_models(model_list):
+            for (model_idx, (model, _model_path)) in model_list:
+                parser = DataParser(self.game, train_data_dir, self.cfg)
+                train_dataset = parser.create_train_dataset()
+
+                keras.backend.set_value(model.optimizer.learning_rate, lr)
+
+                training_start_time = time.time()
+                history = model.fit(train_dataset, epochs=1, verbose=0).history
+                train_dur = time.time() - training_start_time
+
+                values_loss[model_idx] = history["value_head_loss"][0]
+                policy_loss[model_idx] = history["policy_head_loss"][0]
+                value_accuracy[model_idx] = history["value_head_value_head_accuracy"][0]
+                policy_accuracy[model_idx] = history["policy_head_policy_head_accuracy"][0]
+                training_duration[model_idx] = train_dur
+
+                trained_models[model_idx] = (model, self._save_model(model))
+
+        # Divide the training into jobs
+        models = list(enumerate(models))
+        workers_num = min(self.cfg["training"].get("threads", 1), len(models))
+        jobs = []
+        for idx in range(workers_num):
+            i = idx / workers_num * len(models)
+            j = (idx + 1) / workers_num * len(models)
+            jobs.append(models[int(i) : int(j)])
+
+        # Spawn workers_num-1 threads, each executing a single job
+        workers = []
+        if len(jobs) > 1:
+            for models_list in jobs[1:]:
+
+                def worker_main():
+                    train_models(models_list)
+
+                worker = threading.Thread(target=worker_main)

```

Comment:
you can use `functools.partial(train_models, models_list)` instead of `worker_main`